cargo-apk/src/main.rs

```diff
@@ -1,112 +1,96 @@
+    },
+}
+
+#[derive(Parser)]
+struct Args {
+    #[clap(flatten)]
+    subcommand_args: cargo_subcommand::Args,
+    /// Use device with the given serial (see `adb devices`)
+    #[clap(short, long)]
+    device: Option<String>,
+}
+
+#[derive(clap::Subcommand)]
+enum ApkSubCmd {
+    /// Checks that the current package builds without creating an apk
+    Check {
+        #[clap(flatten)]
+        args: Args,
+    },
+    /// Compiles the current package and creates an apk
+    Build {
+        #[clap(flatten)]
+        args: Args,
+    },
+    /// Run a binary or example of the local package
+    Run {
+        #[clap(flatten)]
+        args: Args,
+        /// Do not print `logcat` after running the application
+        #[clap(short, long)]
+        no_logcat: bool,
+    },
+    /// Start a gdb session attached to an adb device with symbols loaded
+    Gdb {
+        #[clap(flatten)]
+        args: Args,
+    },
+    /// Print the version of cargo-apk
+    Version,
+    // TODO:
+    // Test {}
+    // Doc {}
+}
 
 fn main() -> anyhow::Result<()> {
     env_logger::init();
-    let args = std::env::args();
-    let mut device_serial = None;
-    let mut no_logcat = false;
-    let cmd = Subcommand::new(args, "apk", |name, value| match name {
-        "--device" => {
-            if let Some(value) = value {
-                println!("Running on {}", value);
-                device_serial = Some(value.to_owned());
-                Ok(true)
-            } else {
-                Err(cargo_subcommand::Error::InvalidArgs)
-            }
+    let Cmd {
+        apk: ApkCmd::Apk { cmd },
+    } = Cmd::parse();
+    match cmd {
+        ApkSubCmd::Check { args } => {
+            let cmd = Subcommand::new(args.subcommand_args)?;
+            let builder = ApkBuilder::from_subcommand(&cmd, args.device)?;
+            builder.check()?;
         }
-        "--no-logcat" => {
-            no_logcat = true;
-            Ok(true)
-        }
-        _ => Ok(false),
-    })
-    .map_err(Error::Subcommand)?;
-
-    let builder = ApkBuilder::from_subcommand(&cmd, device_serial, no_logcat)?;
-
-    match cmd.cmd() {
-        "check" | "c" => builder.check()?,
-        "build" | "b" => {
+        ApkSubCmd::Build { args } => {
+            let cmd = Subcommand::new(args.subcommand_args)?;
+            let builder = ApkBuilder::from_subcommand(&cmd, args.device)?;
             for artifact in cmd.artifacts() {
                 builder.build(artifact)?;
             }
         }
-        "run" | "r" => {
+        ApkSubCmd::Run { args, no_logcat } => {
+            let cmd = Subcommand::new(args.subcommand_args)?;
+            let builder = ApkBuilder::from_subcommand(&cmd, args.device)?;
             anyhow::ensure!(cmd.artifacts().len() == 1, Error::invalid_args());
-            builder.run(&cmd.artifacts()[0])?;
-        }
-        "--" => {
-            builder.default()?;
+            builder.run(&cmd.artifacts()[0], no_logcat)?;
         }
```

Comment:
Let's see if I can add back support for this, as it was quite useful.

Reply:
It is sort of working, but getting unknown arguments (outside of `cargo_subcommand::Args`) parsed properly is tricky and left out for now.

Reply:
Finally fixed/implemented in https://github.com/rust-windowing/android-ndk-rs/pull/363!